geo/src/algorithm/contains_properly/polygon.rs

```diff
@@ -0,0 +1,563 @@
+        .map(|l| (l, l.bounding_rect()))
+        .any(|(l, l_bbox)| {
+            rhs_bbox_cache
+                .iter()
+                .any(|(r, r_bbox)| l_bbox.intersects(r_bbox) && l.intersects(r))
+        });
+}
+
+/// Given two non-empty polygons with no intersecting boundaries,
+/// Return true if first polygon completely contains second polygon
+fn polygon_polygon_inner_loop<T>(self_poly: &Polygon<T>, rhs_poly: &Polygon<T>) -> bool
+where
+    T: GeoNum,
+{
+    debug_assert!(!self_poly.is_empty() && !rhs_poly.is_empty());
+    debug_assert!(
+        self_poly
+            .lines_iter()
+            .flat_map(|s| rhs_poly.lines_iter().map(move |s2| (s, s2)))
+            .all(|(s, s2)| !s.intersects(&s2))
+    );
+
+    let Some(rhs_ext_coord) = rhs_poly.exterior().0.first() else {
+        return false;
+    };
+
+    if !self_poly.contains_properly(rhs_ext_coord) {
+        // is disjoint
+        return false;
+    }
+
+    //  let Some(lhs_ext_coord) = self_poly.exterior().0.first() else {
+    //         return false;
+    //     };
+
+    //     if rhs_poly.contains_properly(lhs_ext_coord) {
+    //         // is disjoint
+    //         return false;
+    //     }
+
+    // if there exits a self_hole which is not inside a rhs_hole
+    // then there must be some point of rhs which does not lie on the interior of self
+    // and hence self does not contains_properly rhs
+    for self_hole in self_poly.interiors() {
+        // empty hole is always covered
+        let Some(self_hole_first_coord) = self_hole.0.first() else {
+            continue;
+        };
+
+        // hole outside of RHS does not affect intersection
+        if coord_pos_relative_to_ring(*self_hole_first_coord, rhs_poly.exterior())
+            != CoordPos::Inside
+        {
+            continue;
+        }
+
+        // if any RHS point is inside self_hole, then we fail the contains_properly test
+        // since all rings are either concentric or disjoint, we can check using representative point
+        let self_hole_pt_in_rhs_hole = rhs_poly
+            .interiors()
+            .iter()
+            .map(|rhs_ring| coord_pos_relative_to_ring(*self_hole_first_coord, rhs_ring))
+            .any(|pos| pos == CoordPos::Inside);
+        if !self_hole_pt_in_rhs_hole {
+            return false;
+        }
+    }
+    true
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::wkt;
+    use crate::{ContainsProperly, Convert, Relate};
+    use crate::{MultiPolygon, Point, Polygon};
+
+    // basic pairwise test
+    #[test]
+    fn test_poly_for_poly() {
+        let base: Polygon<f64> =
+            wkt! {POLYGON((90 0,90 90,0 90,0 0,90 0),(60 30,60 60,30 60,30 30,60 30))}.convert();
+
+        let donut_inside: Polygon<f64> =
+            wkt! {POLYGON((80 10,80 80,10 80,10 10,80 10),(70 20,70 70,20 70,20 20,70 20))}
+                .convert();
+        let fully_inside: Polygon<f64> = wkt! {POLYGON((20 10,20 20,10 20,10 10,20 10))}.convert();
```

Comment:
Thank you again for writing your tests with WKT! It's so helpful to be able to pop these into JTSTestBuilder.

<img width="300" alt="Image" src="https://github.com/user-attachments/assets/cbfbe309-4b2b-4866-be2b-f154627a763c" />